File: go/p34.go
package main

import "fmt"

func factorial(n int) int {
    product := 1

    for ; n >= 2; n-- {
        product *= n
    }

    return product
}

func factorialDigitSum(n int) int {
    sum := 0

    for n > 0 {
        sum += factorial(n % 10)
        n /= 10
    }

    return sum
}

func main() {
    sum := 0
    for i := 10; i != 2540160; i++ {
        if factorialDigitSum(i) == i {
            sum += i
        }
    }
    fmt.Println(sum)
}

File: cpp/p34.cpp
#include <cassert>
#include <iostream>
using namespace std;

int Factorial(int x)
{
    if (x == 0)
        return 1;
    
    assert(x > 0);
    int product = x;
    while (x != 1)
        product *= --x;
    
    return product;
}

int FactorialDigitSum(int x)
{
    int sum = 0;
    while (x != 0)
    {
        sum += Factorial(x % 10);
        x /= 10;
    }
    
    return sum;
}

int main()
{
    int sum = 0;
    for (int i = 3; i != 2540160; ++i)
        if (FactorialDigitSum(i) == i)
            sum += i;
    
    cout << sum << endl;
    
    return 0;
}
